# **Selection of clonal mutations in tumor tissue for therapy response monitoring in liquid biopsies from cancer patients** 

Currently, response to cancer treatment is monitored by scanning/imaging, which is labor-intensive, costly, and not very sensitive. A more cost-efficient method to assess treatment response would allow for a more fine-grained (more regular) monitoring, resulting in faster detection of relapse and treatment resistance. It is in this context that liquid biopsies come into play, as recent studies have indicated that tumor-derived genetic material can be detected through sequencing in blood and urine from cancer patients, among others.

To make liquid biopsy-based circulating tumor DNA monitoring suitable and cost-effective for therapy response monitoring, the selection of tumor-derived markers to be tracked is the most crucial step. Tumor-specific mutations (absent in normal cells) are ideally suited for this and allow very sensitive and specific detection of tumor material in a normal genetic background. However, not all cells in a tumor have an identical genetic make-up and mutational status. A tumor typically consists of different cell populations (clones) sharing a set of mutations. Some of these mutations are passenger mutations, meaning that they are less important for tumorigenesis and progression, and are typically found in a subset of tumor cells (i.e. a clone). Tracking these would thus mean that the effect of the cancer treatment on only a subset of the tumor cells would be monitored. A better choice are the mutations shared among all tumor cells (including metastases), which are called driver or clonal or truncal mutations (i.e. shared by all tumor clones, part of the trunk in a phylogenetic tree). Assessing such mutations as part of a liquid-based monitoring method would give clinicians a better and broader view of the therapy effectiveness. However, the difficulty lies in separating these clonal mutations from the subclonal (often passenger) mutations based on a tumorâ€™s mutational profile.

## **Dissertation workflow**

### **Aims**
ðŸ“Œ To benchmark various tools for the identification of clonal mutations.  
ðŸ“Œ To identify clonal mutations with the optimal tool using variant-calling format files and CNV information from 30 tumor types as input.

The methods identify clonal mutations based on variant allele frequencies (VAF), leveraging the fact that subclonal variants have lower fractions than clonal mutations in bulk sequencing datasets. Several tools â€” FastClone, PyClone, PhyloWGS, QuantumClone, and SciClone â€” were benchmarked using both simulated and real datasets, including combinations of different sequencing instruments. The optimal tool for this study was selected after evaluating sensitivity, specificity, speed, memory consumption, and overall performance.

QuantumClone emerged as the best-performing tool, allowing me to identify sets of clonal mutations for each tumor type and analyze the presence of driver mutations. Currently, I am working on integrating this tool into an easy-to-use workflow that includes input data preprocessing, formatting, and user-friendly result output in the form of a command-line tool.
